pinc/upload_file.inc

```diff
@@ -0,0 +1,279 @@
+<?php
+include_once($relPath.'misc.inc'); // get_upload_err_msg(), attr_safe(), return_bytes()
+
+// This contains functions for uploading files using a resumable process if
+// supported by the browser. The process takes place in two stages: Showing
+// an upload form, then processing the uploaded data.
+// It is assumed that large uploads will be needed only with zipped files, then
+// the function show_upload_form() can be used. This posts the data to the same
+// php page where the function validate_uploaded_file() can be used.
+// If you need to upload only small files your form can contain an
+// <input type='file'> with name='uploaded_file' Then validate_uploaded_file()
+// can be used to process the upload.
+
+define("RESUMABLE_UPLOAD_SIZE", 1024 * 1024 * 1024);  // 1GB
+
+// If uploading large files this function should be called near the beginning
+// of the page to which the upload is submitted (resumable or not). It detects
+// a problem which can occur with non-resumable uploads when the file uploaded
+// was larger than post_max_size and shows an error instead of failing silently.
+// We do this first because if the POST failed, $_REQUEST and $_POST are empty
+// and we have no data.
+// http://andrewcurioso.com/blog/archive/2010/detecting-file-size-overflow-in-php.html
+function detect_too_large()
+{
+    if($_SERVER['REQUEST_METHOD'] == 'POST' && empty($_POST) &&
+        empty($_FILES) && $_SERVER['CONTENT_LENGTH'] > 0)
+    {
+        die( sprintf(_("Uploaded file is too large. Maximum file size is %s."), get_max_upload_size()));
+    }
+}
+
+function get_max_upload_size()
+{
+    return humanize_bytes(return_bytes(ini_get("upload_max_filesize")));
+}
+
+function get_big_upload_blurb()
+{
+    global $db_requests_email_addr;
+
+    return sprintf(_("If you are trying to upload a very big zip file
+        and the upload does not succeed, upload a small placeholder zip file
+        instead and email %s for assistance."), "<span class='nowrap'>$db_requests_email_addr</span>");
+}
+
+// This supplies the arguments for the header in the page containing the upload
+// form below
+function get_upload_args()
+{
+    global $code_url, $utf8_site;
+
+    $upload_messages = [
+        'working' => _("Working, please wait..."),
```

Comment:
I'm sure there may be many other instances of this, but I prefer to use the single '…' character to get better rendering with fonts. Just minor suggestion

Reply:
Not sure I understand your meaning.

Reply:
instead of using 3 '.' characters, use the single '' unicode character https://www.fileformat.info/info/unicode/char/2026/index.htm to represent ellipsis

Reply:
I think that we tend to use the three separate dots in general because of our proofing guidelines. In fact, that's one of the characters that will continue to be rejected even once we move to UTF-8. The problem is that in a monospace font, the three dots are crammed into the space a single character takes up, so in fact, you don't necessarily get better rendering across fonts.

Reply:
@chrismiceli - Purely from a technical perspective, right now the site is limited to outputting Latin-1 characters (we explicitly set the charset we report to the browser), so we couldn't use the `…` character without using the HTML entity for it. After the Unicode deployment we can start using UTF-8 characters in the code and this becomes possible.

Reply:
> @chrismiceli - Purely from a technical perspective, right now the site is limited to outputting Latin-1 characters (we explicitly set the charset we report to the browser), so we couldn't use the `…` character without using the HTML entity for it. After the Unicode deployment we can start using UTF-8 characters in the code and this becomes possible.

Also true, though I think we will still need to be aware of where the output will be shown as monospace text and where a proportional font will be displayed.

Reply:
I agree entirely with @cpeel and @srjfoo. Not important right now, but something we can consider in the future depending on context.